test/realtime_web/controllers/broadcast_controller_test.exs

```diff
@@ -0,0 +1,128 @@
+      ] do
+        sub_topic_1 = "sub_topic"
+        sub_topic_2 = "sub_topic"
+        topic_1 = tenant.external_id <> ":" <> sub_topic_1
+        topic_2 = tenant.external_id <> ":" <> sub_topic_2
+
+        payload_1 = %{"data" => "data"}
+        payload_2 = %{"data" => "data"}
+
+        conn =
+          post(conn, Routes.broadcast_path(conn, :broadcast), %{
+            "messages" => [
+              %{
+                "topic" => sub_topic_1,
+                "payload" => payload_2
+              },
+              %{
+                "topic" => sub_topic_2,
+                "payload" => payload_2
+              }
+            ]
+          })
+
+        assert_called(Endpoint.broadcast_from(:_, topic_1, "broadcast", payload_1))
+        assert_called(Endpoint.broadcast_from(:_, topic_2, "broadcast", payload_2))
+
+        assert conn.status == 202
+
+        # Wait for counters to increment
+        :timer.sleep(1000)
+        {:ok, rate_counter} = RateCounter.get(Tenants.requests_per_second_key(tenant))
+        assert rate_counter.avg != 0.0
+      end
+    end
+
+    test "returns 422 when batch of messages includes badly formed messages", %{conn: conn} do
+      with_mocks [
+        {Endpoint, [:passthrough], broadcast_from: fn _, _, _, _ -> :ok end}
+      ] do
+        conn =
+          post(conn, Routes.broadcast_path(conn, :broadcast), %{
+            "messages" => [
+              %{
+                "topic" => "topic"
+              },
+              %{
+                "payload" => %{"data" => "data"}
+              },
+              %{
+                "topic" => "topic",
+                "payload" => %{"data" => "data"}
+              }
+            ]
+          })
+
+        assert Jason.decode!(conn.resp_body) == %{
+                 "errors" => %{
+                   "messages" => [
+                     %{"payload" => ["can't be blank"]},
+                     %{"topic" => ["can't be blank"]},
+                     %{}
+                   ]
+                 }
+               }
+
+        assert_not_called(Endpoint.broadcast_from(:_, :_, :_, :_))
+
+        assert conn.status == 422
+      end
+    end
+  end
+
+  describe "unauthorized" do
+    test "invalid token returns 401", %{conn: conn} do
+      tenant = tenant_fixture()
+
+      conn =
+        conn
+        |> put_req_header("accept", "application/json")
+        |> put_req_header("x-api-key", "potato")
+        |> then(&%{&1 | host: "#{tenant.external_id}.supabase.com"})
+
+      conn = post(conn, Routes.broadcast_path(conn, :broadcast), %{})
+      assert conn.status == 401
+    end
+
+    test "invalid tenant returns 401", %{conn: conn} do
+      conn =
+        conn
+        |> put_req_header("accept", "application/json")
+        |> put_req_header("x-api-key", "potato")
+        |> then(&%{&1 | host: "potato.supabase.com"})
+
+      conn = post(conn, Routes.broadcast_path(conn, :broadcast), %{})
+      assert conn.status == 401
+    end
+  end
```

Comment:
Can you add a test for the counter here also?

Reply:
test the counter is not incremented on bad requests?